design/rest-api-documentation.md

```diff
@@ -0,0 +1,133 @@
+# REST API documentation
+
+* **Status**: Notes
+* **JIRA**: 
+  * [KEYCLOAK-9655](https://issues.jboss.org/browse/KEYCLOAK-9655)
+  * [KEYCLOAK-14744](https://issues.jboss.org/browse/KEYCLOAK-14744)
+
+
+## Motivation
+
+The value of an API to a large extent depends on how easy it is to consume. Documentation plays an important part in whether an API is useful. The main issue with the Keycloak documentation these days is to keep it up to date with the changes into the code, forcing developers to dig into the sources sometimes to understand how the API works.
+
+## Microprofile OpenAPI
+
+The Microprofile OpenAPI specification defines a standard mechanism to document REST endpoints using annotations or a pre-generated JSON. That was introduced since *WildFly 19* making things easier to be consumed by [Swagger](https://swagger.io/) tooling. 
+
+To document the external API that's exposed by the [Account REST endpoint](https://github.com/keycloak/keycloak/blob/9369c7cf4d2227ae8be40a6129ca21200ae16a9d/services/src/main/java/org/keycloak/services/resources/account/AccountRestService.java#L464-L517), we need to update the source code:
+
+```
+public class AccountRestService {
+
+    ...
+
+  @Path("/applications")
+  @GET
+  @Produces(MediaType.APPLICATION_JSON)
+  @Operation(description = "List of applications")
+  @APIResponses(value = {
+  	@APIResponse(
+  			responseCode = "200",
+  			description = "Successful, List of applications"
+      ),
+  	@APIResponse(
+  			responseCode = "401",
+  			description = "Unauthorized, Failure"
+      ),
+  	@APIResponse(
+  			responseCode = "500",
+  			description = "Error in getting the list of applications"
+      )
+  })
+  public Response applications(@QueryParam("name") String name) {
+  	...
+  }
+
+}
+```
+
+A new dependency must be introduced in the Keycloak `pom.xml` file:
+
+```
+<dependency>
+    <groupId>org.eclipse.microprofile.openapi</groupId>
+    <artifactId>microprofile-openapi-api</artifactId>
+</dependency>
+```
+
+By spinning up the Keycloak server, developers should be able to send an HTTP request to http://localhost:8080/auth/openapi and get the OpenAPI document. Here is an example:
```

Comment:
Should it not rather be '/auth/openapi/account'? We'll want to have separate OpenAPI documents for different APIs, and eventually we will want this for the admin endpoints (and others) as well.

What about versioning? Would it be '/auth/openapi/account' for the latest, and then you can also grab for a previous version using '/auth/openapi/account/<version>'?

Reply:
Note that Microprofile OpenAPI does not support multiple separate APIs.  The specification dictates that a single OpenAPI file be accessible at the `/openapi` endpoint (in the case of Wildfly it would be `/context/openapi` I assume, although I haven't looked at how Wildfly handles this).  Like all Microprofile specs, the MP-OpenAPI spec was meant to be used in single-application platforms.  So if you have multiple separate APIs in the same application context, and you want to document them as separate OpenAPI files, the Microprofile-OpenAPI spec will not let you do that.

That said, I do think using the MP-OpenAPI annotations as described above is the right thing to do.  It just might be the case that you will need to figure out how to leverage them in a non-standard way).

Reply:
Agree with Eric here. What I had in mind thinking about `admin` vs `account` endpoints make use of tags. For example:

```
@Tag(name="Account REST API", description="Account REST API operations."),
@Tag(name="Admin REST API", description="Admin REST API operations."),
```

I can do some updates into the document to cover these scenarios.